Feature/stability corrected vreman

linneahuusko opened the pull request:
Correction factor that computes the thermal stratification and corrects the eddy viscosity accordingly, for use in ABL simulations. The correction is implemented in the standard Vreman model and is activated using optional keywords in the json file.

linneahuusko committed:
First implementation of stability correction

linneahuusko committed:
First implementation of stability correction

linneahuusko committed:
Fix issue with allocation of gradient fields

linneahuusko committed:
Fix issue with allocation of gradient fields

linneahuusko committed:
Allocate dTdz; add safeguard in computation of correction

linneahuusko committed:
Allocate dTdz; add safeguard in computation of correction

timfelle commented:
> I think the proper workflow is that you can merge whatever you want into a branch you created and in the end when the branch is about to merge into develop, you need to require people to pass the PR. But since this time there are something already be there, we can later merge the [ExtremeFLOW:feature/stability_corrected_vreman](https://github.com/ExtremeFLOW/neko/tree/feature/stability_corrected_vreman) branch into the dev branch without further discussion after this PR is passed.

We could also just change the target to be develop 😉

njansson committed:
Cleanup example makefile (#2167)

Cleanup C API example makefile, remove unused target (we only use C, and
the implicit rules)

Shiyu-Sandy-Du commented on src/les/vreman.f90:
```diff
@@ -73,16 +76,22 @@ subroutine vreman_init(this, fluid, json)
     ! Based on the Smagorinsky Cs = 0.17.
     call json_get_or_default(json, "c", c, 0.07_rp)
     call json_get_or_default(json, "extrapolation", if_ext, .false.)
+    call json_get_or_default(json, "stability_correction", if_corr, .false.)
```
yes that is what I mean. If the buoyancy is pointing upwards, then we don't need this term, do we? We only need it when the buoyancy is pointing down to the wall.

Shiyu-Sandy-Du replied:
ah okay I see!

timofeymukha replied:
buoyancy_correction is good indeed

linneahuusko commented on src/les/vreman.f90:
```diff
@@ -73,16 +76,22 @@ subroutine vreman_init(this, fluid, json)
+    call json_get_or_default(json, "stability_correction", if_corr, .false.)
+    call json_get_or_default(json, "vertical_direction", vertical_dir, "z")
+    call json_get_or_default(json, "ri_c", ri_c, 0.25_rp)
+    call json_get_or_default(json, "theta0", theta0, 293.0_rp)
```
That might be a good idea, although I suppose the correction should work even in a non-dimensional case if an appropriate value for `theta0` is provided?

linneahuusko committed:
stability_correction -> buoyancy_correction

timofeymukha commented on src/les/bcknd/cpu/vreman_cpu.f90:
```diff
@@ -152,6 +157,40 @@ subroutine vreman_compute_cpu(if_ext, t, tstep, coef, nut, delta, c)
     end do
+    if (if_corr .eqv. .true.) then
+          theta => neko_field_registry%get_field_by_name("temperature")
+          call neko_scratch_registry%request_field(dTdz, temp_indices(12))
```
The temp_indices array is of size 9! With index 12, I think you are sending in garbage here, or?

timofeymukha reviewed (commented):
If think you address the comments, the only thing missing for merging to develop is documentation.

timfelle committed:
Introduce matrix version of scratch registry (#2165)

This matrix registry allow temporary matrices to be reused.

When requesting a matrix, the desired size is specified. A new matrix
will be allocated if no matrices of the given size is currently
available.

timfelle committed:
Allow user to set up mesh before load balancing (#2162)

Enable users to configure the mesh prior to load balancing and the
construction of point zones, improving the initialization process.

timfelle committed:
Fix device math test (#2168)

timfelle committed:
Introduce a vector scratch registry (#2159)

This vector registry allow temporary vectors to be reused.

When requesting a vector, the desired size is specified. A new vector
will be allocated if no vectors of the given size is currently
available.

njansson committed:
Add missing link to Filter page in documentation (#2170)

timofeymukha committed:
Add neko_vector_registry (#2155)

Adds a registry for vector_t objects, similar to the field_registry_t.
Has essentially the same capabilities, and has an associated singleton,
neko_vector_registry_t.

I wanted the possibility to retrieve by name, but vectors store no name
unlike fields. So, I added a private named_vector type, which is what is
stored in the registry. However, the user never sees this type, and the
registry returns pointers to raw vector_t.

---------

Co-authored-by: Tim Felle Olsen <timfelle@hotmail.com>

timofeymukha commented:
> One question to all: I could probably write a GPU counterpart of it, but maybe not in a very short future. If it's the case, can we include it into release 1.0 with the lack of GPU implementation?

I guess if it is really-really needed? We have to make a cut-off somewhere, though. I think the ABL users will have to keep track of develop anyway, or? Would it make more sense to keep it out for now, and when we have the k-eq model, and the ABL wall models in, it will be a nice package for the next release?

Shiyu-Sandy-Du commented:
> > > One question to all: I could probably write a GPU counterpart of it, but maybe not in a very short future. If it's the case, can we include it into release 1.0 with the lack of GPU implementation?
> > 
> > 
> > I guess if it is really-really needed? We have to make a cut-off somewhere, though. I think the ABL users will have to keep track of develop anyway, or? Would it make more sense to keep it out for now, and when we have the k-eq model, and the ABL wall models in, it will be a nice package for the next release?
> 
> How about that we have the final cutoff for features end of November (week 48). Then we should have enough time to package things and push up packages (e.g. spack) before the festivities

Considering the ABL users mainly focus on develop for the moment, I think it's not a bad idea to leave buoyancy corrected Vreman for the next PR.

njansson committed:
Add API documentation (#2172)

Adds a baseline documentation of the API, including installation
instructions

Fixes #2171

vbaconnet committed:
Sponge source term (#2139)

njansson committed:
Remove macOS 13 runners from workflow (#2176)

macOS 13 runner will be retired on 251204

njansson committed:
Remove invalid ref to unused gslib (#2175)

njansson committed:
Change Case to use TBP (as all other types) (#2177)

IsaMammadli committed:
Feature/cyclic bc (#2164)

In spectral element method, Direct Stiffness Sumation (DSS) is used to
enforces continuity across element boundary nodes. Neko (as well as
Nek5000) achieves this with gather-scatter operation (`gs%op(..,..,
GS_OP_ADD`). Since the GLL nodes at periodic faces are also _neighbors_
they are subject to GS operation. When the periodic faces do not align
with each other, this summation can no longer be performed in Cartesian
basis (see figure below).
<img width="2352" height="852" alt="cyc"
src="https://github.com/user-attachments/assets/41a35f2f-7177-4593-af46-b0e08f6fda72"
/>

In Nek5000 this is handled with **cyclic boundary conditions**, which
temporarily rotates the vectors at periodic faces to an intermediate
basis to apply DSS correctly. This occurs as following:
_[ rotate forward ] → [ gather ] → [ scatter ] → [ rotate backward ]_

For bent channels, toroidal and helical pipes we rotate the vectors into
Frenet-Serret (a.k.a TNB) basis which is a moving reference frame for
differentiable curves. In this proposed approach, I inserted
`rotate_cyc( ux, uy, uz, idir, coef)` subroutine in relevant locations
of the code, as in the following diagram:
<img width="1299" height="865" alt="cycint"
src="https://github.com/user-attachments/assets/ca322c0c-0190-486d-af2d-0a6310457d41"
/>

- **idir=1** is for Cartesian → TNB rotation and **idir=0** is TNB →
Cartesian.
- The 5 lines of rotate-gs calls are analogous to **opdssum** subroutine
in Nek5000, but kept explicit intentionally since gs%op calls are not
always consecutive in Neko (e.g. in fused_cpld_cg.F90 there are
device_event_sync and blstz%apply in between gs%op calls.)
- A new subroutine `coef_generate_cyclic_bc` computes rotation angle as
part of `coef%init`. It uses surface normals which are also generated in
**coef.f90**
- Here `cyc_msk` array is also generated. It is a mask array that helps
to iterate GLL nodes at periodic faces and components of rotation
matrix.
- CPU and device implementation of rotate_cyc have two versions for rank
1 and rank 4 arrays.
- Cyclic BC is enabled in .case file manually.


There can be different ways to handle these steps, so I am open to
suggestions and I list here some of them for your comments.

1. Do we want to combine rotate - gs process to make it more compact
like opdssum ?
2. Currently rotate_cyc only handles 2D rotations, which are sufficient
for toroids and helical pipes. This is the same as in Nek5000 because
the rotation is geometry dependent and difficult to generalize. However,
it would require minor changes to adjust it for 3D case, if any.
(Unfortunately we dont have an example for that at this point, unless we
orient pipes weirdly.)
3. I think keeping cyclic BC as a manual switch in the user file is a
good idea, because the user must be aware of the requirement, its
implementation details and if current implementation suffices without
any modifications. I have a `check_cyclic` subroutine that hints about
this. We can provide it in the documentation and instruct how
modifications can be done.
4. I don't actually calculate the rotation angle but rather the
components of the 2x2 rotation matrix R11 and R12. These two are
sufficient as R11=R22=cos(theta) and R12=-R21=-sin(theta). So we save 2
arrays [cos(theta) and sin(theta)] instead of 1 angle array [theta], but
avoid calculation of cosine and sine functions every time rotate_cyc is
called. This is the original reason why I followed this approach.
However, in GPU loading 2 arrays from global memory is gonna be slower
than sine-cosine calculation so I think of going back to angles. Let me
know if my reasoning is faulty.
5. I assigned one angle value for every GLL node on periodic faces. This
keeps it more general, if the periodic faces are deformed for some
reason (e.g. a wavy surface).
6. `opr_device_rotate_cyc` GPU implementation does not follow same
approach as other subroutines in **opr_device.f90** i.e. it does not
copy arrays into shared memory before using them. We load and store them
once so I believe shared memory is redundant.
7. Consequently, it is more similar to subroutines in **math.f90** and
**mathops.f90**, so we may consider putting it into another file.
8. Once we clarify these points, we can easily extend implementations to
HIP and OpenCL.


Thanks!

---------

Co-authored-by: Isa Mammadli <b237dc17@alex1.nhr.fau.de>
Co-authored-by: Niclas Jansson <njansson@kth.se>

Shiyu-Sandy-Du committed:
squash the symmetric part of sij in ax_helm_full (#2174)

timfelle committed:
Stabilized-convergence logging (#2153)

Add initial grace period for convergence warnings and checks.

This should help reduce superflous "Solver did not converged" messages
as the flow stabilizes during the initial part of the simulation.

Here we declare a flow "stabilized" on the first iteration for which the
solver converged for all components.

To facilitate programatic use, we also introduce a subroutine to reset
the value.

linneahuusko committed:
Add check to wall model hip backend to avoid crashes (#2178)

When running a simulation on GPU with hip the wall model would crash if
more than one GPU was used. Now there is a check to stop the kernel from
launching if the number of boundary points is zero on a given GPU.

---------

Co-authored-by: Shiyu Du <shiyud@nj.pdc.kth.se>

njansson committed:
Add short best practice guide for performance (#2181)

Fixes #1275

njansson committed:
Fix issue with vector expand using CCE (#2182)

Fixes #2179

njansson committed:
Unify all workflows (remove old release workflow) (#2183)

njansson committed:
Add missing author (#2185)

adampep committed:
Tiny correction of the doxygen description (#2186)

timofeymukha committed:
Add vector kernels to neumann_t (#2188)

njansson committed:
Add missing HIP kernel (#2190)

Shiyu-Sandy-Du committed:
specify the intent of some lately added subroutines in tensor_cpu.f90 (#2189)

tuananhdao committed:
chore: update examples/README.md add missing cases and fix typos (#2191)

## Changes

Update `examples/README.md`

- add missing cases
- fix typos

timfelle committed:
Stop printing empty sections (#2187)

Simple solution to the empty sections in the log.

Starting a section now stores the header in the class.
First message after the section call will print the header and clear the
variable.
Ending a section will also clear the variable.

njansson committed:
Fix implicit OpenMP declaration (#2193)

timofeymukha committed:
User guide for meshing (#2194)

A section on high order meshing should be added later.

---------

Co-authored-by: Tim Felle Olsen <timfelle@hotmail.com>
Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

timfelle committed:
Update our scratch registries (#2173)

Updates to how scratch registries are designed.

- `vector_scratch_registry_t` removed.
- `matrix_scratch_registry_t` removed.

Extended registry
`scratch_registry_t` from which one can request a vector, field or
matrix.
The global registry is this type.

Tests of the scratch registry are now done for pure vector, pure matrix,
pure field and mixed registry with all types.

Shiyu-Sandy-Du committed:
Some debugging and some enhancement for scalar nut (#2180)

1. fix a bug to assign eddy diffusivity field for scalar equation via Pr
2. add a new interface to let scalar scheme to access eddy diffusivity
field calculated from sgs models

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

Shiyu-Sandy-Du committed:
fix a typo of coupled cg residual definition (#2195)

njansson committed:
Fix compilation errors with NAG (#2196)

njansson committed:
Remove unused import (#2198)

njansson committed:
Fix a couple of use without only (#2200)

timfelle committed:
Fix/zeroing_scratch (#2199)

This PR forces users to specify if they want the field requested from a
scratch registry to be `clear`ed before it is passed to them.

This then also influences all places where we used the registry, meaning
we need to be careful that I have not messed up which are returned
zeroed and which are assigned by the user.

Please check carefully, this should be tested

timfelle committed:
Reduce default registry size (#2201)

Decrease the default size of the registry and its expansion size to
optimize resource usage.

timfelle committed:
Unit test on the CI (#2202)

timofeymukha committed:
Fix scalar auto name, throw on duplicate (#2205)

`write(field_name, '(A,I0)') 's_', i` caused a segfault when field_name
is of length zero. It seems, `write` does not automatically reallocate
to the necessary size, unlike assignment. I added a buffer of fixed size
to write to.

I also suggest removing the handling of duplicates and throw an error. I
think a duplicate is a red flag meaning that something went not like the
user wanted.

timfelle committed:
Add missing components and improve field assignment (#2203)

Enhance field assignment by adding missing components and utilizing
`cfill` for scalar assignments. Update tests to verify that names are
correctly copied during assignments.

timofeymukha committed:
Add pysemtools to CI (#2208)

Useful for the python-driven tests.

timfelle committed:
Registry/alias (#2204)

Aliases are no longer arrays.
Rather we use the json as a string to string map function.

njansson committed:
Fix all (except glb_interp) unit tests for OpenCL (#2206)

Co-authored-by: Tim Felle Olsen <timfelle@hotmail.com>

njansson committed:
Add test-suite badge to README (#2211)

Added a badge for the test suite in the README.

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

njansson committed:
Add gitlab ci def (#2213)

njansson committed:
Temporarily disable global interpolation tests for OpenCL backends (#2214)

See #2210

njansson committed:
Update gitlab ci with stages (#2215)

timofeymukha committed:
Add tests for statistics, revert fld writer behaviour, fix scalar_stats_output (#2212)

* Adds a `freeze` component for the scalar_scheme, similar to the fluid.
This is to allow setting synthetic for the scalar in `compute` in the
use file.
* This needed a bit of change in scalar_aux, it treats one KSP only,
passed in a loop from `scalars`, based on the status of `freeze`.
@timfelle the `stabalized` thing was working somehow on all scalars at
the same time, or? I am a bit unsure if I broke something here.
* Adds tests for fluid and scalar *basic* stats. Outputs 1d, 2d and 3d
stats, checks the values against prescribed analytical ones.
* Fixes bug in scalar_stats_output: adds the reshuffling of fields for
consistent output.
* Adds mpi4py to the python environment. Hopefully this does not mess
things up?
* Updates the statistics guide with the new numbering which includes
temperature.

timfelle committed:
Registry/rework (#2207)

This PR significantly reworks the field registry.

- The registry is now renamed to `registry_t` and `neko_registry`
- The registry now supports vectors and matrices as well.
- We should test it more thoroughly in the unit tests now.

timofeymukha committed:
Remove pointwise interfaces from docs (#2216)

njansson committed:
Add integration tests to gitlab runner, and update gold files for cuda (#2218)

njansson committed:
Fix dist target (#2220)

njansson committed:
Update NVIDIA HPC SDK version to 25.11 (#2219)

njansson committed:
Update CI pipeline with OpenCL and HIP tests (#2222)

njansson committed:
Disable CI for hip until we have native runners (#2223)

njansson committed:
Bump minor due to breaking API changes

njansson committed:
Fix subroutine, module scoping issue (#2224)

timofeymukha committed:
Add AGENTS.md (#2226)

AGENTS.md has become a standard for providing instructions to AI, see
agents.md on the web.

I experimented today with Codex, and wrote this file, which only
contains information on how to help users setup a simulation case. I am
generally very happy with the little testing I have done.

- It easily setup a correct case file for a flat plate simulation with
blasius inlet and IC. When I say correct, I mean the case file was valid
and had the stuff I specified baked in. I was not very detailed, so of
course the actual setup is crap (random timestep size and so on).
- Generated the mesh with genmeshbox, given the number of elements in
the boundary layer and min and max spacings inside it (so it ran Python
under the hood to produce the csv that is fed to the mesher)
- Finally, I prompted: "I want my simulation to compute a kinetic energy
field every time step and save it to the hard drive every 10 seconds."
It managed to figure out it needs a user file and that it can use the
field_writer simcomp. The user file was essentially correct, maybe an
extra if statement one could skip, but compiles and does what I asked
for.

So, I think this is a start! I don't know how much we can eventually
rely on AI to write code but I think
- Help with case setup.
- Simple to moderate things in the user file.
- Doing annoying stuff like adding all the necessary things to the build
system for a new test or whatever.
Is totally within reach, and adding AGENT.md files at key locations may
boost this dramatically.

<img width="261" height="193" alt="image"
src="https://github.com/user-attachments/assets/cffd6dd4-af92-4b60-8be6-57edb18dcb2c"
/>

timofeymukha committed:
Add testing section to AGENTS.md, update testing.md (#2230)

tuananhdao committed:
Fix OpenCL memory leak: add missing clReleaseKernel calls (#2227)

njansson committed:
Add missing destructors (#2231)

timofeymukha committed:
Add check for implicit integer conversion in json read (#2232)

json_fortran implicitly converts reals to integers when reading an entry
into an integer.

This adds another read to the same entry with a real variable and checks
that no rounding has occurred when populating the integer.

Also adds a skeleton for unit testing json_utils. Currently, just check
that integer reads do work when json contains reals that have a
0 decimal.

njansson committed:
Add missing end_section (#2234)

timofeymukha committed:
Fix mean_output_t initilization (#2235)

Asked AI why it thinks we get this weird error with user_stats.csv
failing to appear randomly in the pytest CI, and it directly found this
very sad bug.

`this%start_time` was not initialized, so randomly the check of time >
this%start_time would fail...

We should not have let that error slide like that.

timfelle committed:
Fix HDF5 bug (#2236)

Address a bug in the HDF5 handling related to field assignment by
ensuring the name is only assigned when it is empty.

nick-morse committed:
Fix kernel launch failures for ranks without cyclic boundaries (#2237)

tuananhdao committed:
Entropy viscosity for the compressible solver (#2068)

## Changes

This PR implements "entropy viscosity" for the compressible module.
There has been some literature which proves that the resulting solution
is entropy-consistent. And ultimately, the solution converges to the
entropy solution.

The main idea is
- To use the entropy residual as a distributor of the artificial
viscosity
- No accuracy is lost when solution is smooth (entropy residual is
essentially zero)
- Entropy residual tracks the shocks/discontinuities
- The method adds just enough viscosity to stabilize the solution,
guided by local wave speeds
- The CFL number doesn't change, totally depending on the advective flux
- The parameters `c_avisc_low=0.5` and `c_avisc_entropy=1.0` should work
for all cases, unless user wants to fine-tune them to make the solution
sharper/smoother.

**Reference**: Guermond, J. L., Pasquetti, R., & Popov, B. (2011).
Entropy viscosity method for nonlinear conservation laws. Journal of
Computational Physics, 230(11), 4248-4267.
[[Link](https://www.sciencedirect.com/science/article/pii/S0021999110006583)]

## Use

Insert `c_avisc_low` and `c_avisc_entropy` to the `numerics` block. If
not set, `c_avisc_low=0.5` and `c_avisc_entropy=1.0` are applied by
default.

```json
        "numerics": {
            "time_order": 3,
            "polynomial_order": 5,
+           "c_avisc_low": 0.5,
+           "c_avisc_entropy": 1.0
        }
```

## Supported backends

(ticked means initial tests done)

- [x] CPU
- [x] CUDA
- [x] HIP
- [x] OpenCL

## To-do list

- [x] Test strong scaling. Known scaling issue from Niclas: "too much
MPI traffic per node that would eat up the memory". Will be fixed in
another PR.
- [x] Add documentation about the compressible solver

## Simulation Results

<p align="center"><b>Before and after runs use the SAME MESH and
POLYNOMIAL ORDER</b></p>
<p align="center">The settings below are chosen so that they run <5 min
on a personal laptop using 4 threads</p>

| Before | After |
| ------ | ----- |
| Forward-facing step ~1K elements, polynomial_order=5 | Forward-facing
step ~1K elements, polynomial_order=5 |
|
![LO](https://github.com/user-attachments/assets/c222c6e9-c8a2-46ee-a7c8-64814974d7e8)
|
![EV](https://github.com/user-attachments/assets/d9a02e4b-31cf-42ca-9b4c-6b4806b19b34)
|
| Cylinder 6939 elements, polynomial_order=5 | Cylinder 6939 elements,
polynomial_order=5 |
| <img width="2454" height="1532" alt="cylinder_before"
src="https://github.com/user-attachments/assets/fcc20693-16e7-45c7-a811-5972f7770a49"
/> | <img width="2454" height="1532" alt="cylinder_after"
src="https://github.com/user-attachments/assets/0999550e-b4c2-4ff0-a395-a1871096d7fd"
/> |
| TGV 10x10x10 mesh, polynomial_order=5, t=0.2 | TGV 10x10x10 mesh,
polynomial_order=5, t=0.2 |
| <img width="2454" height="1532" alt="tgv_before"
src="https://github.com/user-attachments/assets/400a8242-9cc4-41a6-aed2-ea7dc26ae2f9"
/> | <img width="2454" height="1532" alt="tgv_after"
src="https://github.com/user-attachments/assets/7e0854c3-b108-408a-9030-34eed4e7f035"
/> |

njansson committed:
Correctly allocate flux array, based on init flux (#2239)

fixes #2238

njansson committed:
Reduce data movement and launch latency in compressible scheme (#2240)

This PR refactors the compressible euler's step function. The transition
from `field_math` to backend specific kernels were necessary to allow
for better vectorisation (CPU) and reduced kernel launch latency
(accelerators). Data movement has also been reduce for all backends.

njansson committed:
Fix various memory leaks (#2241)

Fix various missing cleanups, but also force deallocation of scratch
arrays instead of assuming (hoping) it will be removed in due time

njansson committed:
Fix memory leak in scratch registry (#2242)

timofeymukha committed:
Add missing distclean (#2243)

Causes make distclean to throw an error.

Happy new year, everyone :-).

njansson committed:
Fix various regression and issues in the C API (#2249)

This fixes various issues and regressions in the C API

- Fix issue with GNU compilers (fixes #2104, #2247)

- Fix regression in pyneko (and Neko.jl) (fixes #2248)

njansson committed:
Fixed regression in neko C-API header (#2250)

njansson committed:
Update copyright year to 2026

njansson committed:
Add step to maximize build space in workflow (#2252)

timofeymukha committed:
Code style improvements (#2245)

This PR is sponsored by your favorite AI overlord.

On a serious note, now that an LLM agent can fix these things for you
based on the flint report while you get coffee, I see no reason not to
make a push for 10.0 in the repo.

Some of the edited files here are not quite bumped to 10.0 yet, but all
should be improved. I stopped at some point so that the PR doesn't grow
way to big.

---------

Co-authored-by: Tim Felle Olsen <timfelle@hotmail.com>

timofeymukha committed:
Code style improvements 2 (#2246)

Dep #2245

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

timofeymukha committed:
Adds 2 recent papers to the readme (#2258)

timofeymukha committed:
Fix scalar boundary conditions (#2253)

This fixes two issues

- The scalar scheme did not implement the boundary condition gs_op min
/max fix for handling nodes touching the boundary with an edge. This PR
adds this treatment.

- Since boundary conditions are applied twice as per above, the
bc_t%updated flag was introduced before, and the fluid scheme takes care
of reseting that properly, but the scalar scheme did not. So e.g. `user`
boundary conditions did not work properly for the scalar, they were only
applied once at the beginning of the simulation. The latter is actually
a typical use case, so not caught before. The reseting is now properly
handled.

njansson committed:
Refactor gs dervied type (#2259)

Refactor the gs derived types, adding pre-computed block offsets in the
base gs type.

This will allow for concurrent gs_ops on CPUs (similar to how the device
backend is processed)

njansson committed:
Fix index range in concurrent loop for gs_scatter_kernel (#2260)

This should fix the recent regression test fails

timofeymukha committed:
Code style improvements 3 (#2256)

- Had to tweak flinter options about max argument and local number. I
guess those are not needed really.
- Didn't have the heart to run the formatter on the axhelm cpu files,
since things are meticulously manually aligned there with the line
breaks. Not sure what to do with this, I would maybe force-merge to
preserve the alignment.

timfelle committed:
Missing free in output types (#2264)

timfelle committed:
Fix missing deallocations (#2262)

timfelle committed:
Introduce deprecation logging (#2261)

Enhance logging functionality by introducing a deprecation logging
feature, updating log levels, and providing documentation on usage.

This allow us to mark a feature as deprecated, while optionally give an
expected removal version and some extra info of how to update the code.

As an example, the use of the operator `ortho` with an implicit device
array, is marked as deprecated and a new `device_ortho` function is
added to replace it.

The logging is bound to new logging levels, which is documented in the
`appendices.md`. With an option to throw an error on call, generating a
stack trace with most compilers to make it easier to figure out where
the code was called.

timfelle committed:
Deprecate implicit device array ortho operator (#2267)

Deprecate the `ortho` operator for implicit device arrays in favor of
`device_ortho`. Update logging to inform users of the deprecation and
provide examples of the new usage. Clean up related code and
documentation.

timfelle committed:
Allow neko to initialize with explicit filename (#2263)

Neko can now be initialized with an explicit filename, providing more
flexibility in case file handling. If no filename is provided, it
defaults to using command line arguments.

Shiyu-Sandy-Du committed:
add api for scratch registry init and free (#2266)

timfelle committed:
Fix checkpoint initialization and improve scalar lag handling (#2265)

Enhance the initialization of scalar lags in the checkpoint and improve
the handling of scalar lag terms. Address readability issues in device
copy operations and resolve various challenges related to checkpoint
management.

timofeymukha committed:
Code style improvements 4 (#2257)

Dep #2256

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>
Co-authored-by: Tim Felle Olsen <timfelle@hotmail.com>

timofeymukha committed:
Add scalars to registry, add registered data to case file (#2228)

The background here is that I want to enable a smoother way of defining
values for case setup entries that are shared between various parts of
the case setup. Concrete examples: there are several things that depend
on `g`, the gravity vector, and on some `U0`, a freestream velocity.
Currently, you have to set `g` or `U0 `in each of those case objects. If
you want to change the value, you have to also change it everywhere and
make sure that things are in sync. As cases get more complex, this may
lead to user errors.

My solution is to introduce a new entry to the case file, which looks
like this:
```json
        "constants":
        [
                {
                    "name": "c1",
                    "value": 1.0
                },
                {
                    "name": "int1",
                    "value": 1
                },
                {
                    "name": "g2",
                    "value": [1, 0, 0.0]
                }
          ]
```
The naming here are very much up to discussion. In particular, scalars
get two separate meanings in the case file here, because it is also used
for scalar solvers... See also below.

Given this, the user can use these registered entries in other parts of
the case setup, e.g.
```json
            "source_terms": [
                {
                    "type": "boussinesq",
                    "scalar_field": "temperature",
                    //"g": [ 0, 0, 1 ], can still set directly, of course
                    "g": "g2", // Point to g2 in the constants
                    "reference_value": 0,
                    "beta": 1
                }
            ]
```
So, by referencing the same entry in the registered_data, we can have a
single point of definition for a parameter across case components.

### Under the hood
The constants parsed by the constructor of `case_t` early. What it does
is put these entries into `neko_const_registry`, which is aa separate
registry just for this purpose. For arrays, I use the vector_t registry
entries, but for scalars I now expanded the registry to also accept
scalars. Here, again, there is the issue of using the word "scalar".
Although it fits very naturally in the registry scalar -> vector ->
matrix, so I think here it is less of an issue than in the case file.
Regardless, suggestions welcome.

So, an added bonus of this system is that you can add things to the
registry from the case file to later use in the user file.

I add a submodule to json_utils is introduced to parse case file entries
and retrieve them either directly or via the registry. I use submodule
so that json_utils itself remains clean of dependencies (`registry_t`
itself uses `json_utils`, so it is important). For integers, we still
use the real scalars in the registry, but check that no rounding has
occurred.

### Misc
- Assignment routine to vector_t from an array of reals.
- A routine to pretty-print the registry contents.

---------

Co-authored-by: Tim Felle Olsen <timfelle@hotmail.com>
Co-authored-by: Victor Baconnet <baconnet@kth.se>
Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

Shiyu-Sandy-Du committed:
Increase length of SOURCE_KNOWN_TYPES elements (#2269)

make sure gjp is fully included in the error message

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

linneahuusko committed:
Fix index for temp_indices

timofeymukha committed:
Throw when params are of wrong type in _default JSON routines (#2270)

The _or_default routines for json parsing used the default value when
the parameter is found in the JSON, but the type is wrong. Since this
situation pretty much must be caused by a user error, we now throw a
neko_error instead.

GrahamBenHarper committed:
Documentation: add some brief comments to Poisson example (#2273)

linneahuusko committed:
Rewrite for arbitrary vertical direction

linneahuusko committed:
Merge branch 'feature/stability_corrected_vreman' of github.com:linneahuusko/neko into feature/stability_corrected_vreman

linneahuusko committed:
Merge branch 'develop' into feature/stability_corrected_vreman

linneahuusko committed:
Update `request_field` call

linneahuusko committed:
Bug fix and safety checks

linneahuusko committed:
Make indices of `aij` terms consistent

linneahuusko committed:
Add default value for g vector

linneahuusko committed:
Add error message if buoyancy correction is called on

linneahuusko committed:
Documentation of buoyancy correction

linneahuusko committed:
Remove duplicate `use` call
